File: fotoapparat/src/test/java/io/fotoapparat/parameter/selector/SelectorsTest.java
package io.fotoapparat.parameter.selector;

import android.support.annotation.Nullable;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentMatchers;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Collection;
import java.util.List;

import static io.fotoapparat.test.TestUtils.asSet;
import static java.util.Arrays.asList;
import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertNull;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.verify;

@RunWith(MockitoJUnitRunner.class)
public class SelectorsTest {

    @Mock
    SelectorFunction<Collection<String>, String> functionA;
    @Mock
    SelectorFunction<Collection<String>, String> functionB;

    @Test
    public void firstAvailable() throws Exception {
        // Given
        List<String> options = asList("B", "C");

        given(functionA.select(options))
                .willReturn(null);

        given(functionB.select(options))
                .willReturn("B");

        // When
        String result = Selectors
                .firstAvailable(
                        functionA,
                        functionB
                )
                .select(options);

        // Then
        assertEquals("B", result);
    }

    @Test
    public void filtered() throws Exception {
        // Given
        SelectorFunction<Collection<String>, String> function = Selectors.filtered(
                functionA,
                new Predicate<String>() {
                    @Override
                    public boolean condition(@Nullable String value) {
                        return value != null && value.startsWith("A");
                    }
                }
        );

        given(functionA.select(ArgumentMatchers.<String>anyCollection()))
                .willReturn("A");

        // When
        String result = function.select(asList("A", "B", "AB"));

        // Then
        assertEquals("A", result);

        verify(functionA).select(asSet("A", "AB"));
    }

    @Test
    public void single() throws Exception {
        // When
        String result = Selectors
                .single("b")
                .select(asList(
                        "a", "b", "c"
                ));

        // Then
        assertEquals("b", result);
    }

    @Test
    public void single_NotFound() throws Exception {
        // When
        String result = Selectors
                .single("b")
                .select(asList(
                        "a", "c"
                ));

        // Then
        assertEquals(null, result);
    }

    @Test
    public void nothing() throws Exception {
        // When
        String result = Selectors.<Collection<String>, String>nothing()
                .select(asList("A", "B", "C"));

        // Then
        assertNull(result);
    }
}

File: fotoapparat/src/test/java/io/fotoapparat/selector/SelectorsTest.kt
package io.fotoapparat.selector

import org.junit.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class SelectorsTest {

    @Test
    fun `Select first available`() {
        // Given
        val options = listOf("B", "C")

        val selectorA: Iterable<String>.() -> String? = { null }
        val selectorB: Iterable<String>.() -> String? = { "B" }

        // When
        val result = firstAvailable(
                selectorA,
                selectorB
        )(options)

        // Then
        assertEquals(
                expected = "B",
                actual = result
        )
    }

    @Test
    fun `Filter selections`() {
        // Given
        lateinit var receivedStrings: Iterable<String>
        val predicate: (String) -> Boolean = {
            it.startsWith("A")
        }
        val selector: Iterable<String>.() -> String? = {
            receivedStrings = this
            "A"
        }

        // When
        val result = filtered(
                selector,
                predicate
        )(listOf(
                "A",
                "B",
                "AB"
        ))

        // Then
        assertEquals(
                expected = "A",
                actual = result
        )
        assertEquals(
                expected = listOf("A", "AB"),
                actual = receivedStrings
        )
    }

    @Test
    fun `Select single which is in the list`() {
        // When
        val result = single(
                preference = "b"
        )(listOf(
                "a",
                "b",
                "c"
        ))

        // Then
        assertEquals(
                expected = "b",
                actual = result
        )
    }

    @Test
    fun `Select single which is not in the list`() {
        // When
        val result = single(
                preference = "b"
        )(listOf(
                "a",
                "c"
        ))

        // Then
        assertEquals(
                expected = null,
                actual = result
        )
    }

    @Test
    fun `Select nothing`() {
        // When
        val result = nothing<String>()(listOf("A", "B", "C"))

        // Then
        assertNull(result)
    }

    @Test
    fun `Select highest value`() {
        // When
        val result = highest<Int>()(listOf(10, 20, 30))

        // Then
        assertEquals(
                expected = 30,
                actual = result
        )
    }

    @Test
    fun `Select highest value, but list was empty`() {
        // When
        val result = highest<Int>()(emptyList())

        // Then
        assertNull(result)
    }

    @Test
    fun `Select lowest value`() {
        // When
        val result = lowest<Int>()(listOf(10, 20, 30))

        // Then
        assertEquals(
                expected = 10,
                actual = result
        )
    }

    @Test
    fun `Select lowest value, but list was empty`() {
        // When
        val result = lowest<Int>()(emptyList())

        // Then
        assertNull(result)
    }

}